pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -31,10 +41,82 @@ import (
 	"istio.io/istio/pkg/log"
 )
 
-// Endpoints implements MeshDiscovery.Endpoints()
-func Endpoints(ds *v1.DiscoveryService, serviceClusters []string) *xdsapi.DiscoveryResponse {
+// EDS returns the list of endpoints (IP:port and in future labels) associated with a real
+// service or a subset of a service, selected using labels.
+//
+// The source of info is a list of service registries.
+//
+// Primary event is an endpoint creation/deletion. Once the event is fired, EDS needs to
+// find the list of services associated with the endpoint.
+//
+// In case of k8s, Endpoints event is fired when the endpoints are added to service - typically
+// after readiness check. At that point we have the 'real' Service. The Endpoint includes a list
+// of port numbers and names.
+//
+// For the subset case, the Pod referenced in the Endpoint must be looked up, and pod checked
+// for labels.
+//
+// In addition, ExternalEndpoint includes IPs and labels directly and can be directly processed.
+//
+// TODO: for selector-less services (mesh expansion), skip pod processing
+// TODO: optimize the code path for ExternalEndpoint, no additional processing needed
+// TODO: if a service doesn't have split traffic - we can also skip pod and lable processing
+// TODO: efficient label processing. In alpha3, the destination policies are set per service, so
+// we may only need to search in a small list.
+
+var (
+	// It doesn't appear the logging framework has support for per-component logging, yet
+	// we need to be able to debug EDS2.
+	edsDebug = len(os.Getenv("PILOT_DEBUG_EDS")) == 0
+
+	edsClusterMutex sync.Mutex
+	edsClusters     = map[string]*EdsCluster{}
```

Comment:
It'll almost certainly be better (faster, less contention) to use a `sync.Map` here rather than a mutex-guarded map since this mutex will be relatively contested and each EdsCluster has a mutex guarding its fields anyway (so you should never need to hold a lock on this map to modify parts of an element inside of it, since the element itself has another lock).

Reply:
I highly doubt contention rate is an issue for few 1000 endpoints changing with the rate workloads get started/stopped. Even a global mutex would hold at this load. 

Of course it can be optimized before we go to 1M endpoints.